src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs

```diff
@@ -22,35 +21,48 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.C
     [Export(typeof(IWorkspaceProjectContextFactory))]
     internal partial class CPSProjectFactory : IWorkspaceProjectContextFactory
     {
+        private readonly IThreadingContext _threadingContext;
         private readonly VisualStudioProjectFactory _projectFactory;
         private readonly VisualStudioWorkspaceImpl _workspace;
         private readonly IProjectCodeModelFactory _projectCodeModelFactory;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public CPSProjectFactory(
+            IThreadingContext threadingContext,
             VisualStudioProjectFactory projectFactory,
             VisualStudioWorkspaceImpl workspace,
             IProjectCodeModelFactory projectCodeModelFactory)
         {
+            _threadingContext = threadingContext;
             _projectFactory = projectFactory;
             _workspace = workspace;
             _projectCodeModelFactory = projectCodeModelFactory;
         }
 
-        IWorkspaceProjectContext IWorkspaceProjectContextFactory.CreateProjectContext(
+        IWorkspaceProjectContext IWorkspaceProjectContextFactory.CreateProjectContext(string languageName, string projectUniqueName, string projectFilePath, Guid projectGuid, object hierarchy, string binOutputPath)
+        {
+            return _threadingContext.JoinableTaskFactory.Run(async () =>
+                await ((IWorkspaceProjectContextFactory)this).CreateProjectContextAsync(
+                    languageName, projectUniqueName, projectFilePath, projectGuid, hierarchy, binOutputPath, CancellationToken.None).ConfigureAwait(false));
+        }
+
+        async Task<IWorkspaceProjectContext> IWorkspaceProjectContextFactory.CreateProjectContextAsync(
             string languageName,
             string projectUniqueName,
             string projectFilePath,
             Guid projectGuid,
             object hierarchy,
-            string binOutputPath)
+            string binOutputPath,
+            CancellationToken cancellationToken)
         {
-            var visualStudioProject = CreateVisualStudioProject(languageName, projectUniqueName, projectFilePath, hierarchy as IVsHierarchy, projectGuid);
+            var visualStudioProject = await CreateVisualStudioProjectAsync(
+                languageName, projectUniqueName, projectFilePath, hierarchy as IVsHierarchy, projectGuid, cancellationToken).ConfigureAwait(false);
             return new CPSProject(visualStudioProject, _workspace, _projectCodeModelFactory, projectGuid, binOutputPath);
```

Comment:
This is another potential case where somebody better not check the cancellationToken prior to the return new.... To be honest I don't see why the separate private method is needed here -- it only seems to increase risk of a bug like this.

Reply:
yup.  thanks.  done.